File: 2444-longest-ideal-subsequence/longest-ideal-subsequence.java
class Solution {
    public int longestIdealString(String s, int k) {
        int n=s.length();
        int[][] dp=new int[n][27];
        for(int i=0;i<n;i++){
            Arrays.fill(dp[i],-1);
        }
        return fun(-1,0,s,k,dp);
    }

    int fun(int prev,int curr,String s,int k,int[][] dp){

        if(curr==s.length()){
            return 0;
        }

        if(dp[curr][prev+1]!=-1){
            return dp[curr][prev+1];
        }

        int not_taken=fun(prev,curr+1,s,k,dp);
        int taken=0;
        if(prev==-1||Math.abs(prev-(s.charAt(curr)-'a'))<=k){
            taken=1+fun(s.charAt(curr)-'a',curr+1,s,k,dp);
        }
        return dp[curr][prev+1]=Math.max(taken,not_taken);
    }
}

File: 2444-longest-ideal-subsequence/longest-ideal-subsequence.cpp
class Solution {
    using ll=long long;
public:
    int longestIdealString(string s, int k) {
        int n=s.size();
        vector<vector<ll>>dp(n,vector<ll>(27,-1));
        return fun(0,-1,s,k,dp);
    }

    ll fun(int ind,int prev,string &s,int k,vector<vector<ll>>&dp){

        if(ind==s.size()){
            return 0;
        }

        if(dp[ind][prev+1]!=-1){
            return dp[ind][prev+1];
        }

        ll not_taken=fun(ind+1,prev,s,k,dp);
        ll taken=0;
        if(prev==-1 || abs((s[ind]-'a')-prev)<=k){
            taken=1+fun(ind+1,s[ind]-'a',s,k,dp);
        }
        return dp[ind][prev+1]=max(taken,not_taken);

    }
};